### ISSUE TITLE
[Bug]: Get unsupported type err when run mo if bf16 is set by set_output_type

### ISSUE THREAD

**Issue by LinGeLin** [openvinotoolkit/openvino#25190]
### OpenVINO Version

2024.0.0

### Operating System

Ubuntu 20.04 (LTS)

### Device used for inference

CPU

### Framework

ONNX

### Issue description

When bf16 is set by set_output_type during the development of a user-defined operator, an error occurs when mo is run

```
[ ERROR ]  -------------------------------------------------
[ ERROR ]  ----------------- INTERNAL ERROR ----------------
[ ERROR ]  Unexpected exception happened.
[ ERROR ]  Please contact Model Optimizer developers and forward the following information:
[ ERROR ]  Check 'error_message.empty()' failed at src/frontends/onnx/frontend/src/frontend.cpp:122:
FrontEnd API failed with GeneralFailure:
Errors during ONNX translation:
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_13/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2173 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_12/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_5/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_9/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2173':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_7/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1862 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_12/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_5/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1862':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1934>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_12/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2171 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_11/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_4/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_8/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2171':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_6/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1864 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_11/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_4/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1864':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1786>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_11/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2177 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_10/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_3/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_7/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2177':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_5/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1866 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_10/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_3/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1866':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1749>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_10/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2179 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_9/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_2/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_6/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2179':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_4/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1868 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_9/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_2/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1868':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1564>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_9/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2175 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_8/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_1/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_5/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2175':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_3/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1870 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_8/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_1/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1870':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1638>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_8/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2181 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_7/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_4/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2181':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_2/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1872 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_7/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1872':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1897>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_15/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1860 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_13/Reshape[0]:f32[100,1,4], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_6/Reshape[0]:bf16[100,1,299]) -> (dynamic[...])' with friendly_name 'Concat_1860':
Argument element types are inconsistent.

[ ERROR ]  Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/openvino/tools/mo/convert_impl.py", line 889, in _convert
    ov_model, legacy_path = driver(argv, {"conversion_parameters": non_default_params})
  File "/usr/local/lib/python3.10/dist-packages/openvino/tools/mo/convert_impl.py", line 552, in driver
    graph, ngraph_function = prepare_ir(argv)
  File "/usr/local/lib/python3.10/dist-packages/openvino/tools/mo/convert_impl.py", line 406, in prepare_ir
    ngraph_function = moc_pipeline(argv, moc_front_end)
  File "/usr/local/lib/python3.10/dist-packages/openvino/tools/mo/moc_frontend/pipeline.py", line 285, in moc_pipeline
    ov_model = moc_front_end.convert(input_model)
  File "/opt/intel/openvino/python/openvino/frontend/frontend.py", line 18, in convert
    converted_model = super().convert(model)
openvino._pyopenvino.GeneralFailure: Check 'error_message.empty()' failed at src/frontends/onnx/frontend/src/frontend.cpp:122:
FrontEnd API failed with GeneralFailure:
Errors during ONNX translation:
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_13/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2173 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_12/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_5/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_9/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2173':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_7/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1862 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_12/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_5/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1862':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1934>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_12/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2171 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_11/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_4/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_8/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2171':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_6/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1864 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_11/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_4/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1864':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1786>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_11/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2177 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_10/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_3/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_7/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2177':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_5/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1866 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_10/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_3/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1866':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1749>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_10/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2179 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_9/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_2/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_6/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2179':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_4/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1868 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_9/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_2/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1868':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1564>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_9/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2175 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_8/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_1/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_5/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2175':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_3/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1870 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_8/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_1/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1870':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1638>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_8/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_2181 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_7/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape/Reshape[0]:bf16[100,1,4784], opset1::Unsqueeze StatefulPartitionedCall/model_1/tf.expand_dims_4/ExpandDims[0]:f32[...]) -> (dynamic[...])' with friendly_name 'Concat_2181':
Argument element types are inconsistent.
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_2/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1872 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_7/Reshape[0]:f32[100,1,128], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape/Reshape[0]:bf16[100,1,4784]) -> (dynamic[...])' with friendly_name 'Concat_1872':
Argument element types are inconsistent.
While validating ONNX node '<Node(Gemm): Gemm__1897>':
Exception from op/constant.hpp:151:
unsupported type
While validating ONNX node '<Node(Concat): StatefulPartitionedCall/model_1/tf.concat_15/concat>':
Check 'element::Type::merge(inputs_et, inputs_et, get_input_element_type(i))' failed at src/core/src/op/concat.cpp:36:
While validating node 'opset1::Concat Concat_1860 (opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_13/Reshape[0]:f32[100,1,4], opset1::Reshape StatefulPartitionedCall/model_1/tf.reshape_6/Reshape[0]:bf16[100,1,299]) -> (dynamic[...])' with friendly_name 'Concat_1860':
Argument element types are inconsistent.
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [X] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by rkazants** [comment]
Hi @LinGeLin, please provide more details:
1. cmd-line for conversion
2. implementation of user-defined operator. Are you sure that bf16 is needed for user-defined operator? user-defined operator should return exactly the same type as in the original framework model.

Also, we ask our user to transition to `ovc` tool instead of `mo`. `mo` is legacy and deprecated path: https://docs.openvino.ai/2024/documentation/openvino-extensibility.html

@slyalin, @kblaszczak-intel, btw, I didn't find a good documentation to register extensions via `ovc`. Here it still refers Model Optimizer: https://docs.openvino.ai/2024/documentation/openvino-extensibility.html

Best regards,
Roman

**Comment by LinGeLin** [comment]
Sorry. It's probably my own problem. Based on the error message, I found that the Concat Op reported this error during the validate_and_infer_types function call. https://github.com/openvinotoolkit/openvino/blob/master/src/core/src/op/concat.cpp#L36

The output of my custom operator, along with the partial input of the model, serves as input to the Concat Op. But these inputs to the model are of type fp32. So the above error occurs when I set the output type of the custom operator to bf16.
